In [1]:
#
import pandas as pd
pd.set_option('display.max_columns', None)

#
import numpy as np

#
import warnings

## 01. Exploración de datos.

In [2]:
df_flight_activity = pd.read_csv("../data/Customer Flight Activity.csv")
df_flight_activity.shape

(405624, 10)

In [3]:
df_loyalty = pd.read_csv("../data/Customer Loyalty History.csv")
df_loyalty.shape

(16737, 16)

### DF1: Flight Activity

In [4]:
df_flight_activity.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0
2,100140,2017,1,6,0,6,1200,120.0,0,0
3,100214,2017,1,0,0,0,0,0.0,0,0
4,100272,2017,1,0,0,0,0,0.0,0,0


In [5]:
df_flight_activity.tail()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
405619,999902,2018,12,0,0,0,0,0.0,0,0
405620,999911,2018,12,0,0,0,0,0.0,0,0
405621,999940,2018,12,3,0,3,1233,123.0,0,0
405622,999982,2018,12,0,0,0,0,0.0,0,0
405623,999986,2018,12,0,0,0,0,0.0,0,0


In [6]:
df_flight_activity.info()

<class 'pandas.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


In [7]:
# No negative numbers
(df_flight_activity < 0).any().any()

np.False_

In [8]:
# Year Verification: the unique values correspond to calendar years
df_flight_activity['Year'].unique()

array([2017, 2018])

In [9]:
# Month verification: numbers must be between 1 and 12
df_flight_activity['Month'].between(1, 12, inclusive='both').all()

np.True_

*DataFrame Info Observations:*

- No negative values detected in the DF.
- No null values detected in the DF.
- Column names follow a consistent style: they include spaces and use Title Case.
- "Year" and "Month" are stored as int instead of datetime. Numbers correspond to year and month values.
    * Decision: Maintained as int since the dataset lacks full date strings (missing days). Keeping them as integers ensures lighter performance and smoother DataFrame operations.
    * ⬜ Check columns with high zero counts to verify if they contain other values.
- Significant difference in total entries between DataFrames: he same customer can appear multiple times across different flights.
    * ⬜ Count the number of unique customers in the dataset.
    * ⬜ Check for overall duplicate records in the DataFrame.
- Reviewed the study requirements and analyzed the available columns in the DataFrame: 
    * ⬜ Assess whether "Points Redeemed" and "Dollar Cost" are useful features.

In [10]:
df_flight_activity['Loyalty Number'].unique().shape

(16737,)

In [11]:
df_flight_activity.duplicated().sum()

np.int64(1864)

In [12]:
df_flight_activity['Flights with Companions'].unique().shape

(12,)

In [13]:
df_flight_activity['Points Redeemed'].unique().shape

(587,)

In [14]:
df_flight_activity['Dollar Cost Points Redeemed'].unique().shape

(49,)

- ✅ Count the number of unique customers in the dataset:
    * Update: Validated; the unique customer count matches the rows in secondary DataFrame.
- ✅ Check for overall duplicate records in the DataFrame:
    * Update: Duplicates verified. 
- ✅ Check columns with high zero counts to verify if they contain other values:
    * Update: Verified; all columns contain unique values other than zero.
- ✅ Assess whether 'Points Redeemed' and 'Dollar Cost' are useful features.
    * Update: Evaluated both columns; although they contain valid non-null entries, they are non-informative for the scope of this analysis. Therefore, a decision was made to drop both features.

- Final Data Cleaning Checklist:
    * ⬜ Remove the duplicate records from the DataFrame.
    * ⬜ Drop 'Points Redeemed' and 'Dollar Cost' columns from the DataFrame.

In [15]:
df_flight_activity.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0
2,100140,2017,1,6,0,6,1200,120.0,0,0
3,100214,2017,1,0,0,0,0,0.0,0,0
4,100272,2017,1,0,0,0,0,0.0,0,0


### DF2: Loyalty History

In [16]:
df_loyalty.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


In [17]:
df_loyalty.info()

<class 'pandas.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  str    
 2   Province            16737 non-null  str    
 3   City                16737 non-null  str    
 4   Postal Code         16737 non-null  str    
 5   Gender              16737 non-null  str    
 6   Education           16737 non-null  str    
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  str    
 9   Loyalty Card        16737 non-null  str    
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  str    
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: float64(4), 

In [18]:
df_loyalty_numeric = df_loyalty.select_dtypes(include='number')
(df_loyalty_numeric < 0).any().any()

np.True_

In [19]:
df_nulos_loyalty = round(df_loyalty.isnull().sum()/df_loyalty.shape[0]*100, 2)
df_nulos_loyalty = df_nulos_loyalty[df_nulos_loyalty>0]
df_nulos_loyalty

Salary                25.32
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64

*DataFrame Info Observations:*

- Negative values detected:
    * ⬜ Check the numerical columns to verify where are the negative values.
- Column names follow a consistent style: they include spaces and use Title Case. Also consistent with the other DataFrame.
    * ⬜ Verify the number of unique customers in the dataset.
- The "Salary" column presents a significant missing data rate of 25%.
    * ⬜ Add to the null value handling pipeline.
- "Cancellation Year" and "Cancellation Month" exhibit a high missingness rate of 87.65%. This indicates they represent active customers, meaning null imputation is not required. Additionally, the columns are stored as float instead of int, because of the presence of null values.
    * ⬜ Assess whether these columns are useful features for the scope of the study.

In [20]:
# No negative numbers in Loyalty Number
df_loyalty['Loyalty Number'].min()

np.int64(100018)

In [21]:
# Negative numbers detected in Salary
df_loyalty['Salary'].min()

np.float64(-58486.0)

In [22]:
# No negative numbers in CLV
df_loyalty['CLV'].min()

np.float64(1898.01)

In [23]:
# Enrollment Year verification: No negatives. The unique values correspond to calendar years.
df_loyalty['Enrollment Year'].unique()

array([2016, 2014, 2013, 2012, 2015, 2018, 2017])

In [24]:
# Enrollment Month verification: numbers must be between 1 and 12
df_loyalty['Enrollment Month'].between(1, 12, inclusive='both').all()

np.True_

In [25]:
# Cancellation Year verification: No negatives. The unique values correspond to calendar years.
df_loyalty['Cancellation Year'].dropna().unique()

array([2018., 2015., 2017., 2014., 2016., 2013.])

In [26]:
# Cancellation Month verification: numbers must be between 1 and 12
df_loyalty['Cancellation Month'].dropna().between(1, 12, inclusive='both').all()

np.True_

In [27]:
# Verification that the customer count matches the flight DataFrame
df_loyalty['Loyalty Number'].value_counts().sum()

np.int64(16737)

In [28]:
df_loyalty.duplicated().sum()

np.int64(0)

* ✅ Check the numerical columns to verify where are the negative values:
    - Update: negative values detected in Salary. 
* ✅ Verify the number of unique customers in the dataset:
    - Update: Confirmed; zero duplicate entries found.
* ✅ Assess whether these columns are useful features for the scope of the study:
    - Update: For the purpose of this study, the 'Cancellation Year' and 'Cancellation Month' columns are not required. However, a new feature will be created to preserve the customer's active status.

- Final Data Cleaning Checklist:
    * ⬜ Clean the negative numbers from 'Salary' column.
    * ⬜ Create new columns with customer's status.
    * ⬜ Drop the columns 'Cancellation Year' and 'Cancellation Month' from the DataFrame.
    * ⬜ Handle null values in the 'Salary' column.

In [29]:
df_loyalty.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


### Merging the DataFrames

In [30]:
df_merge = df_flight_activity.merge(df_loyalty, on= ['Loyalty Number'], how = 'right')

In [31]:
df_merge.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,2017,1,0,0,0,0,0.0,0,0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,480934,2017,2,3,0,3,2823,282.0,0,0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
2,480934,2017,3,0,0,0,0,0.0,0,0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
3,480934,2017,4,0,0,0,0,0.0,0,0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
4,480934,2018,10,6,2,8,3352,335.0,465,38,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN


### Data Cleaning

**Data Cleaning Checklist:**

* ⬜ Create new columns with customer's status.
* ⬜ Drop 'Points Redeemed', 'Dollar Cost', 'Cancellation Year' and 'Cancellation Month' columns from the DataFrame.
* ⬜ Remove the duplicate records from the DataFrame.
* ⬜ Clean the negative numbers from 'Salary' column.
* ⬜ Handle null values in the 'Salary' column.

In [32]:
# Create a new column indicating whether a customer is active or not by handling the null values.
df_merge['Customer Active'] = df_merge['Cancellation Year'].isnull()

In [33]:
# Dropping columns that will not be used in the study.
df_merge.drop(columns=['Dollar Cost Points Redeemed','Points Redeemed','Cancellation Year','Cancellation Month'], inplace=True)

In [34]:
df_merge.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Customer Active
0,480934,2017,1,0,0,0,0,0.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
1,480934,2017,2,3,0,3,2823,282.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
2,480934,2017,3,0,0,0,0,0.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
3,480934,2017,4,0,0,0,0,0.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
4,480934,2018,10,6,2,8,3352,335.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True


In [ ]:
df_merge.duplicated().sum()

np.int64(1864)

In [ ]:
# Dropping duplicated rows
df_merge.drop_duplicates(inplace=True)

In [37]:
df_merge.duplicated().sum()

np.int64(0)